gobblin-runtime/src/main/java/org/apache/gobblin/runtime/spec_catalog/FlowCatalog.java

```diff
@@ -364,6 +364,64 @@ public Spec getSpecWrapper(URI uri) {
    * @return a map of listeners and their {@link AddSpecResponse}s
    */
   public Map<String, AddSpecResponse> put(Spec spec, boolean triggerListener) throws Throwable {
+    Map<String, AddSpecResponse> responseMap = updateHelper(spec, triggerListener);
+    AddSpecResponse<String> compileResponse = responseMap.get(ServiceConfigKeys.COMPILATION_RESPONSE);
+    // Check that the flow configuration is valid and matches to a corresponding edge
+    if (isCompileSuccessful(compileResponse.getValue())) {
+      FlowSpec flowSpec = (FlowSpec) spec;
+      Object syncObject = getSyncObject(flowSpec.getUri().toString());
+      synchronized (syncObject) {
+        try {
+          if (!flowSpec.isExplain()) {
+            long startTime = System.currentTimeMillis();
+            specStore.addSpec(spec);
+            metrics.updatePutSpecTime(startTime);
+          }
+          responseMap.put(ServiceConfigKeys.COMPILATION_SUCCESSFUL, new AddSpecResponse<>("true"));
+        } catch (IOException e) {
+          throw new RuntimeException("Cannot add Spec to Spec store: " + flowSpec, e);
+        } finally {
+          syncObject.notifyAll();
+          this.specSyncObjects.remove(flowSpec.getUri().toString());
+        }
+      }
+    } else {
+      responseMap.put(ServiceConfigKeys.COMPILATION_SUCCESSFUL, new AddSpecResponse<>("false"));
+    }
+
+    return responseMap;
+  }
```

Comment:
This seems very similar to update besides that it calls specStore.addSpec(spec) vs specStore.updateSpec(spec, version). Is it possible for us to parameterize this or move logic around? e.g. `validateFlowCompilation(Flowspec spec, Callable<Void> specStoreCallable)` and send in a as a parameter. It's only used twice though so current still works, but this helps with repetition.